src/group/managed_group/mod.rs

```diff
@@ -0,0 +1,276 @@
+use crate::ciphersuite::*;
+use crate::codec::*;
+use crate::creds::*;
+use crate::errors::ConfigError;
+use crate::framing::{sender::*, *};
+use crate::group::*;
+use crate::key_packages::*;
+use crate::messages::{proposals::*, *};
+use crate::tree::node::*;
+
+/// A `ManagedGroup` represents an `MLSGroup` with an easier, high-level API designed to be used in production. The API
+/// exposes high level functions to manage a group by adding/removing members, get the current member list, etc.
+///
+/// The API is modelled such that it can serve as a direct interface to the Delivery Serive. Functions that modify the public state
+/// of the group will return a `Vec<MLSMessage>` that can be sent to the Delivery Service directly. Conversely, incoming messages
+/// from the Delivery Service can be fed into `process_nessage()`.
+///
+/// A `ManagedGroup` has an internal queue of pending proposals that builds up as new messages are processed. When creating proposals,
+/// those messages are not automatically appended to this queue, instead they have be processed again through `process_message()`.
+/// This allows the Delivery Service to reject them (e.g. if they reference the wrong epoch).
+///
+/// If incoming messages or applied operations are semantically or syntactily incorrect, the function will return a corresponding
+/// error message and the state of the group will remain unchanged.
+///
+/// The application policy for the group can be enforced by implementing the validator callback functions and selectively allowing/
+/// disallowing each operation (see `ManagedGroupCallbacks`)
+///
+/// Changes to the group state are dispatched as events through callback functions (see ManagedGroupCallbacks).
+pub struct ManagedGroup {
+    managed_group_config: ManagedGroupConfig,
+    group: MlsGroup,
+    proposal_queue: ProposalQueue,
+    own_kpbs: Vec<KeyPackageBundle>,
+}
+
+impl ManagedGroup {
+    // === Group creation ===
+
+    /// Creates a new group from scratch with only the creator as a member.
+    pub fn new(
+        credential_bundle: &CredentialBundle,
+        managed_group_config: &ManagedGroupConfig,
+        group_id: GroupId,
```

Comment:
Could we choose the `group_id` randomly in here instead?

Reply:
Since group IDs have to be globally unique they should not be chosen by clients, otherwise you'd get conflicts.